### PR TITLE
#162949082 Favorite and Unfavorite Articles

### PR DESCRIPTION
**Pull request by johnwayodi**
### What does this PR do?
Adds the ability for a user to favorite and unfavorite articles.
A user will also be able to view a list of all his/her favorite articles

### Description of Task to be completed?
- [x] Add tests
- [x] Create Favorites application module
- [x] Create model for Favorite relation
- [x] Add Favorite serializers
- [x] Create Favorite views
- [x] Have the following endpoints working:
     - GET `api/v1/favorites/`
     - PUT `api/v1/articles/<article_slug>/favorite`
     - DELETE `api/v1/articles/<article_slug>/favorite`
- [x] Add extra favorite fields to article response

### How should this be manually tested?
1. Clone this repository on your local machine using 

     `git clone https://github.com/andela/ah-robotics.git`
2. Change your working directory to the project's directory using 
     
     `cd ah-robotics`
3. Change to the specific branch `ft-favorite-articles-162949082` using 

     `git checkout ft-favorite-articles-162949082`
4. Create a virtual environment using the following command: 
     
     `python3 -m virtualenv env`
5. Activate the virtual environment using 
     
     `source env/bin/activate`
6. Install all the requirements using 
     
     `pip install -r requirements.txt`

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
        `pg_ctl -D /usr/local/var/postgres start`
    
    2. Run psql interactive terminal
    
        `psql postgres`
    
    3. Create a new user:
    
        `CREATE USER sample_username WITH PASSWORD 'sample_password';`
    
    4. Grant privileges to the user:
    
        `ALTER USER sample_username CREATEDB;`
    
    5. Create a database:
    
        `CREATE DATABASE sample_database_name WITH OWNER sample_username;`
    
    6. Exit psql interactive with the following command: `\q`
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get cloudinary values, create an account with cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      `source .env`
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      `python manage.py makemigrations`
      
      (ii) Apply the migrations using: 
      
      `python manage.py migrate`

10. Run the following command to collect static files.

    `python manage.py collectstatic`
      
#### Testing the features

1. Run the server from the terminal using: 
    
    `python manage.py runserver`
2. Ensure [Postman](https://www.getpostman.com/) is installed, open it on your local machine and Paste the URL `http://localhost:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and the add the following payload to the body:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```
5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Log in the user using the following endpoint `http://localhost:8000/api/v1/users/login/` and insert the following payload to the body section.
    
    ```
        {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```
    Upon successful login, a token is returned to the user, 
    this token will be used to access the protected endpoints
7. To Create an article, insert the following URL: `http://localhost:8000/api/v1/articles/`. In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, post an article by using the following payload:
    
    ```
    {
        "article": {
            "title": "Sample Article",
            "description": "Sample Article Description",
            "body": "The sample article body",
            "tagList": ["samples", "articles"],
        }
    }
    ```
    Once all the fields are inserted, change request type to a **POST** request and click the **Send** button.

8. To add an article to favorites, add the article's slug to the url and then send a **PUT** request to the URL `http://localhost:8000/api/v1/articles/<article_slug>/favorite`.

    > example request:
    >
    > `http://localhost:8000/api/v1/articles/sample-article/favorite`.
    
9. To see whether the article has been added to favorites, head over to `http://localhost:8000/api/v1/articles/` and perform a **GET** request, you should see that the article *favorited* value is now changed to **true** and the *favorite_count* has been **incremented** by one.

10. To view a list of articles that the current user has added to favorites, 
send a **Get** request to the url `http://localhost:8000/api/v1/favorites/`

11. To remove an article from favorites, add the slug of the article that you wish to remove from favorites
    
    example `http://localhost:8000/api/v1/articles/sample-article/favorite`
    
    and then change the request type to **DELETE** and click send 

12. On loading the articles list again in the url `http://localhost:8000/api/v1/articles/`, you'll see that the article is no longer favorited and the count is reduced by 1.
#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
The **PUT** request method can also be used in place of **POST** in some instances to represent creation of a resource.
In this task, in oder to get addition and removal of favorites in the same view, the `RetrieveUpdateDestroyAPIView` was used as it allows both PUT and DELETE request types on the same URL resource.

### What are the relevant pivotal tracker stories?
[#162949082](https://www.pivotaltracker.com/n/projects/2235171/stories/162949082)

### Screenshots
1. Create an article
    screen shot 2019-01-30 at 11 56 07 pm

2. Adding article to favorites
    screen shot 2019-01-30 at 11 24 04 pm

4. List of favorite articles by the user
    screen shot 2019-01-30 at 11 24 44 pm

5. Article view after added to favorites
    screen shot 2019-01-30 at 11 25 12 pm

6. Remove an article from favorites
    screen shot 2019-01-30 at 11 25 45 pm

7. Article view after removing from favorites
    screen shot 2019-01-30 at 11 26 16 pm

## Questions:
Questions and comments are welcome.

### CODE REVIEW

File: authors/apps/favorites/tests/base_test.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        self.article = {
+            "article": {
+                "title": "The One",
+                "description": "This is an article about the one",
+                "body": "The one is the one",
+                "tagList": ["sample", "article"],
+                "author": 1}
+        }
+        # self.article_slug = {
```
Comment (jomasim) [inline authors/apps/favorites/tests/base_test.py]: remove the commented code

Comment (johnwayodi) [reply authors/apps/favorites/tests/base_test.py]: @jomasim I have removed the commented code. thanks for reviewing

File: authors/apps/favorites/tests/base_test.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                "tagList": ["sample", "article"],
+                "author": 1}
+        }
+        self.article_slug = "the-one"
+        self.non_existent_article_slug = "article-absent"
+        self.client = APIClient()
+        self.register_url = reverse('authentication:register_url')
+        self.login_url = reverse('authentication:login_url')
+        self.articles_url = reverse('articles:articles')
+        # self.favorite_url = reverse('favorites:favorites')
```
Comment (jomasim) [inline authors/apps/favorites/tests/base_test.py]: Remove this commented line

Comment (johnwayodi) [reply authors/apps/favorites/tests/base_test.py]: @jomasim, I have removed the commented code. thanks again for reviewing

Comment (jomasim) [reply authors/apps/favorites/tests/base_test.py]: @johnwayodi cool otherwise good work I have successfully tested the feature locally.

**Comment by Teatoller** [comment]
Hi @johnwayodi 

Kindly on *Testing the features* point no ```7```

**add a coma** on the payload on the line with the _body_. 
The payload sample is causing an error.


screenshot 2019-01-31 at 12 46 10

Regards
Steven Ennis

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @johnwayodi ,
I have followed the steps, tested and everything works well from my side.

**Comment by Teatoller** [comment]
Hi @johnwayodi 

With changes discussed, I have been able to test successfully

Regards
Steven Ennis

File: authors/apps/articles/serializers.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
...
 
-from authors.apps.profiles.serializers import AuthorSerializer
-from authors.apps.profiles.models import UserProfile
 from taggit_serializer.serializers import (
     TagListSerializerField, TaggitSerializer)
 
+from authors.apps.profiles.models import UserProfile
+from authors.apps.profiles.serializers import AuthorSerializer
+from authors.apps.favorites.models import Favorite
+
```
Comment (sharkdevs) [inline authors/apps/articles/serializers.py]: remove this blank line. let all your local imports should be together

**Review by sharkdevs — COMMENTED** [review]
Works fine. Consider the following feedback

File: authors/apps/articles/serializers.py (hunk 3 above)
Comment (johnwayodi) [reply authors/apps/articles/serializers.py]: @sharkdevs , thanks for reviewing. I have made the requested changes.

**Review by waracci — COMMENTED** [review]
Works as expected! 👍

**Pull request merged**